# The Task
The datasets are simplified versions of data at fictional company called Bon. Analyze the data and present your results to senior management:

* Summarizes sales and operations - think about which statistics, metrics, or figures are important to our business

* Propose 2-3 areas where the company can improve

* Any other analysis you find crucial to include and highlight certain business KPIs

In [51]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

In [52]:
customer = pd.read_csv('https://raw.githubusercontent.com/ShopBonsai/Data-Analyst-Test/main/data-analyst-test/customer_data.csv')
order = pd.read_csv('https://raw.githubusercontent.com/ShopBonsai/Data-Analyst-Test/main/data-analyst-test/order_data.csv')
order_prod = pd.read_csv('https://raw.githubusercontent.com/ShopBonsai/Data-Analyst-Test/main/data-analyst-test/order_product_data.csv')
prod = pd.read_csv('https://raw.githubusercontent.com/ShopBonsai/Data-Analyst-Test/main/data-analyst-test/product_data.csv')

In [53]:
# since left-most column is unnamed column, skip the left-most one
customer = customer.iloc[:,1:]
order = order.iloc[:,1:]
order_prod = order_prod.iloc[:,1:]
prod = prod.iloc[:,1:]

# Exploratory Data Analysis

To do deeper analytics, we need to understand the shape of the data. 

From EDA, 3 things we want to answer are:
    
* Examine columns and see how we can merge dataframes 
* Identify how many null values (missing values) we have for each dataframe
* Perform summary statistics for any dataframe that is applicable (ones that have numerical values)

In [54]:
order.head()

order_id                 customer_order_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status         purchased_at          approved_at         delivered_at  
0    delivered  2017-10-02 10:56:33  2017-10-02 11:07:15  2017-10-10 21:25:13  
1    delivered  2018-07-24 20:41:37  2018-07-26 03:24:27  2018-08-07 15:27:45  
2    delivered  2018-08-08 08:38:49  2018-08-08 08:55:23  2018-08-17 18:06:29  
3    delivered  2017-11-18 19:28:06  2017-11-18 19:45:59  2017-12-02 00:28:42  
4    delivered  2018-02-13 21:18:39  2018-02-13 22:20:29  2018-02-16 18:17:02

From below, it looks like there are many different types of orderings

In [55]:
order.order_status.unique()

array(['delivered', 'invoiced', 'shipped', 'processing', 'unavailable',
       'canceled', 'created', 'approved'], dtype=object)

In [56]:
order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   order_id           99441 non-null  object
 1   customer_order_id  99441 non-null  object
 2   order_status       99441 non-null  object
 3   purchased_at       99441 non-null  object
 4   approved_at        99281 non-null  object
 5   delivered_at       96476 non-null  object
dtypes: object(6)
memory usage: 4.6+ MB


In [57]:
order.isnull().sum()

order_id                0
customer_order_id       0
order_status            0
purchased_at            0
approved_at           160
delivered_at         2965
dtype: int64

In [58]:
customer.head()

customer_order_id                customer_unique_id  zipcode
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0    14409
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3     9790
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e     1151
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c     8775
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066    13056

In [59]:
customer.isnull().sum()

customer_order_id     0
customer_unique_id    0
zipcode               0
dtype: int64

In [60]:
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_order_id   99441 non-null  object
 1   customer_unique_id  99441 non-null  object
 2   zipcode             99441 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 2.3+ MB


In [61]:
order_prod.head()

order_id  num_items_in_order  \
0  00010242fe8c5a6d1ba2dd792cb16214                   1   
1  00018f77f2f0320c557190d7a144bdd3                   1   
2  000229ec398224ef6ca0657da4fc703e                   1   
3  00024acbcdf0a6daa1e931b038114c75                   1   
4  00042b26cf59d7ce69dfabb4e55b4fd9                   1   

                         product_id   price  shipping  
0  4244733e06e7ecb4970a6e2683c13e61   58.90     13.29  
1  e5f2d52b802189ee658865ca93d83a8f  239.90     19.93  
2  c777355d18b72b67abbeef9df44fd0fd  199.00     17.87  
3  7634da152a4610f1595efa32f14722fc   12.99     12.79  
4  ac6c3623068f30de03045865e4e10089  199.90     18.14

In [62]:
order_prod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 5 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   order_id            112650 non-null  object 
 1   num_items_in_order  112650 non-null  int64  
 2   product_id          112650 non-null  object 
 3   price               112650 non-null  float64
 4   shipping            112650 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 4.3+ MB


In [63]:
order_prod.isnull().sum()

order_id              0
num_items_in_order    0
product_id            0
price                 0
shipping              0
dtype: int64

In [64]:
order_prod.describe()

num_items_in_order          price       shipping
count       112650.000000  112650.000000  112650.000000
mean             1.197834     120.653739      19.990320
std              0.705124     183.633928      15.806405
min              1.000000       0.850000       0.000000
25%              1.000000      39.900000      13.080000
50%              1.000000      74.990000      16.260000
75%              1.000000     134.900000      21.150000
max             21.000000    6735.000000     409.680000

In [65]:
prod.head()

product_id  number_of_photos   category
0  1e9e8ef04dbcff4541ed26657ea517e5                 1  perfumery
1  6a2fb4dd53d2cdb88e0432f1284a004c                 2  perfumery
2  0d009643171aee696f4733340bc2fdd0                 1  perfumery
3  b1eae565a61935e0011ee7682fef9dc9                 2  perfumery
4  8da90b37f0fb171b4877c124f965b1f6                 3  perfumery

In [66]:
prod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32328 entries, 0 to 32327
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   product_id        32328 non-null  object
 1   number_of_photos  32328 non-null  int64 
 2   category          32328 non-null  object
dtypes: int64(1), object(2)
memory usage: 757.8+ KB


In [67]:
prod.isnull().sum()

product_id          0
number_of_photos    0
category            0
dtype: int64

In [68]:
prod.describe()

number_of_photos
count      32328.000000
mean           2.188815
std            1.736746
min            1.000000
25%            1.000000
50%            1.000000
75%            3.000000
max           20.000000

From below, it looks like there are multiple product categories. I can see there are opportunities for product breakdown and segmentation later on.

In [69]:
prod.category.unique()

array(['perfumery', 'art', 'sports', 'baby', 'home', 'music',
       'miscellaneous', 'furniture', 'toys', 'construction_tools',
       'computers_accessories', 'health_beauty', 'luggage_accessories',
       'garden_tools', 'office_furniture', 'auto', 'electronics',
       'fashion_shoes', 'phones', 'stationery', 'fashion_accessories',
       'computers', 'watches', 'pet_shop', 'industrials',
       'air_conditioning', 'games', 'books', 'costruction_tools_tools',
       'fashion_underwear_beach', 'fashion_male', 'party_supplies',
       'food_drink', 'costruction_tools_garden', 'fashion_female',
       'audio', 'fashion_sport', 'christmas', 'fashion_kids',
       'photography', 'flowers'], dtype=object)

After reviewing dataframes, we discovered that there are no need to impute or remove null values. We know that there are null values in delivered_at from order dataframe but this is just the way it is - some products are not delivered yet and hence the field is just empty.

## Order analytics

Not that total item value, total shipping cost, total profit (total item value - total shipping cost). This will give us overview of which category of products is most profitable.

#### Top 5 categories in terms of # of items sold, avg. profit, sum profit

In [70]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov['profit'] = aov['price'] - aov['shipping']
aov = aov.groupby(['category'])['profit'].size().reset_index()
aov.columns = ['category', 'n_items_sold']
aov.sort_values(by='n_items_sold', ascending = False).head(5)
#aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
#aov.columns = ['category', 'avg_order_item_value', 'avg_shipping_cost', 'avg_profit']
#aov.sort_values(by = 'avg_order_item_value', ascending = False)

category  n_items_sold
22              furniture         20022
26                   home          9566
25          health_beauty          9465
37                 sports          8431
8   computers_accessories          7644

In [71]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov['profit'] = aov['price'] - aov['shipping']
aov = aov.groupby(['category'])['profit'].mean().reset_index()
aov.columns = ['category', 'avg_profit']
aov.sort_values(by='avg_profit', ascending = False).head(5)
#aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
#aov.columns = ['category', 'avg_order_item_value', 'avg_shipping_cost', 'avg_profit']
#aov.sort_values(by = 'avg_order_item_value', ascending = False)

category   avg_profit
7          computers  1050.344322
30             music   229.854123
27       industrials   203.497197
40           watches   182.287223
0   air_conditioning   161.912007

In [72]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov['profit'] = aov['price'] - aov['shipping']
aov = aov.groupby(['category'])['profit'].sum().reset_index()
aov.columns = ['category', 'sum_profit']
aov.sort_values(by='sum_profit', ascending = False).head(5)

category  sum_profit
22      furniture  1466299.43
40        watches  1068020.84
25  health_beauty  1054173.91
26           home   963982.09
37         sports   791448.19

#### summary statistics of price/shipping cost/profit by category

In [73]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov['profit'] = aov['price'] - aov['shipping']
aov = aov.groupby(['category'])['price','shipping','profit'].agg(['sum','mean','count']).reset_index()
#aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
#aov.columns = ['category', 'avg_order_item_value', 'avg_shipping_cost', 'avg_profit']
#aov.sort_values(by = 'avg_order_item_value', ascending = False)
aov.sort_values(by = 'category')

<ipython-input-73-22519996ba9f>:9: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



category       price                       shipping  \
                                     sum         mean  count        sum   
0           air_conditioning    53323.56   184.510588    289    6530.99   
1                        art    25471.31   115.254796    221    4198.36   
2                      audio    50570.60   139.697790    362    5684.72   
3                       auto   578966.65   139.847017   4140   90488.10   
4                       baby   401922.63   133.131047   3019   66851.21   
5                      books    68423.24    79.840420    857   13811.31   
6                  christmas     8737.84    58.252267    150    3190.91   
7                  computers   218684.14  1098.915276    199    9665.62   
8      computers_accessories   888724.61   116.264339   7644  143999.16   
9         construction_tools   220838.71   157.741936   1400   31494.46   
10  costruction_tools_garden    25096.28   108.173621    232    5187.66   
11   costruction_tools_tools    15903.95   154.407282    103    2030.22   
12               electronics   155043.93    56.813459   2729   45679.16   
13       fashion_accessories   149329.39    75.228912   1985   30558.90   
14            fashion_female     2634.94    58.554222     45     583.50   
15              fashion_kids      519.95    74.278571      7      78.72   
16              fashion_male    10452.33    83.618640    125    2062.62   
17             fashion_shoes    23001.46    89.499844    257    4813.37   
18             fashion_sport     2094.52    72.224828     29     563.03   
19   fashion_underwear_beach     9305.95    73.275197    127    1854.07   
20                   flowers     1110.04    33.637576     33     488.87   
21                food_drink    67258.86    58.844147   1143   17233.24   
22                 furniture  1871620.31    93.478189  20022  405320.88   
23                     games   148003.37   135.907594   1089   19094.36   
24              garden_tools   470495.28   110.237882   4268   96650.40   
25             health_beauty  1233131.72   130.283330   9465  178957.81   
26                      home  1171935.09   122.510463   9566  207953.00   
27               industrials   109064.91   231.560318    471   13217.73   
28       luggage_accessories   138701.63   128.785172   1077   30076.49   
29             miscellaneous   659598.21   156.228851   4222   93314.16   
30                     music   195622.68   256.050628    764   20014.13   
31          office_furniture   268154.31   160.763975   1668   67057.05   
32            party_supplies     4465.19   106.314048     42     820.19   
33                 perfumery   390144.65   116.809775   3340   53026.98   
34                  pet_shop   211695.64   110.028919   1924   38918.56   
35                    phones   365175.44    77.945665   4685   73843.72   
36               photography     6674.47    95.349571     70    1227.71   
37                    sports   954852.55   113.254958   8431  163404.36   
38                stationery   231317.10    90.748176   2549   47012.56   
39                      toys   471286.48   116.944536   4030   75774.58   
40                   watches  1166176.98   199.040276   5859   98156.14   

                          profit                      
         mean  count         sum         mean  count  
0   22.598581    289    46792.57   161.912007    289  
1   18.997104    221    21272.95    96.257692    221  
2   15.703646    362    44885.88   123.994144    362  
3   21.857029   4140   488478.55   117.989988   4140  
4   22.143495   3019   335071.42   110.987552   3019  
5   16.115881    857    54611.93    63.724539    857  
6   21.272733    150     5546.93    36.979533    150  
7   48.570955    199   209018.52  1050.344322    199  
8   18.838195   7644   744725.45    97.426145   7644  
9   22.496043   1400   189344.25   135.245893   1400  
10  22.360603    232    19908.62    85.813017    232  
11  19.710874    103    13873.73   134.696408    103  
12  16.738424   2729   1

In [74]:
graph = aov.reset_index()[['category','profit']]

In [75]:
graph_clean = pd.concat([graph['category'],graph['profit'][['sum']]],axis = 1)

In [76]:
graph_clean.columns = ['category', 'total sum of profit']

In [77]:
fig = px.bar(graph_clean, x = 'category', y = 'total sum of profit')


#fig.update_layout(uniformtext_minsize=3, uniformtext_mode='hide')
fig.update_traces(textfont_size=10)
fig.show()

#### Average order value (AOV) by month

In [78]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov['profit'] = aov['price'] - aov['shipping']

aov = aov.groupby(['purchased_month','order_id'])['price','shipping','profit'].sum().reset_index()
aov = aov.groupby(['purchased_month'])['price','shipping','profit'].mean().reset_index()

aov.columns = ['purchased_month', 'avg_order_value', 'avg_shipping_cost', 'avg_profit']

aov = pd.melt(aov, id_vars=["purchased_month"], value_vars=["avg_order_value", "avg_shipping_cost", "avg_profit"])

aov.columns = ['purchased_month', 'type', 'value in $']

aov['purchased_month'] = aov['purchased_month'].astype('str')


#aov.set_index('purchased_month').plot()
fig = px.line(aov, 
        x = 'purchased_month', 
        y = 'value in $', 
        color = 'type', 
        text = round(aov['value in $'],0) )

fig.update_traces(textfont_size=10, textposition='top center')

fig.update_layout(
    height=500,
    title_text='average order value/shipping cost/profit per order sold by month'
)

<ipython-input-78-8f892ea40778>:10: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.

<ipython-input-78-8f892ea40778>:11: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



#### Average order item value (AOIV) by month

In [79]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov['profit'] = aov['price'] - aov['shipping']

aov = aov.groupby(['purchased_month'])['price','shipping','profit'].mean().reset_index()

aov['purchased_month'] = aov['purchased_month'].astype('str')
#aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
aov.columns = ['purchased_month', 'avg_order_item_value', 'avg_shipping_cost', 'avg_profit']

aov = pd.melt(aov, id_vars=["purchased_month"], value_vars=["avg_order_item_value", "avg_shipping_cost", "avg_profit"])

aov.columns = ['purchased_month', 'type', 'value in $']


#aov.set_index('purchased_month').plot()
fig = px.line(aov, 
        x = 'purchased_month', 
        y = 'value in $', 
        color = 'type', 
        text = round(aov['value in $'],0) )

fig.update_traces(textfont_size=10, textposition='top center')

fig.update_layout(
    height=500,
    title_text='average value/shipping cost/profit per item sold by month'
)

<ipython-input-79-158d1cd4dd4f>:10: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



In [80]:
# aov = order.merge(order_prod, on = 'order_id')
# aov = aov[aov.order_status == 'delivered']
# aov = aov.merge(prod, on = 'product_id')

# per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

# aov['purchased_month'] = per
# aov['profit'] = aov['price'] - aov['shipping']
# aov = aov.groupby(['purchased_month','category'])['price','shipping','profit'].sum().reset_index()

# #aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
# aov.columns = ['purchased_month', 'category', 'total_order_item_value', 'total_shipping_cost', 'total_profit (000s)']
# aov = aov.sort_values(by = ['purchased_month','total_profit (000s)'], ascending = False)
# aov = aov.groupby(['purchased_month']).head(4)
# aov['purchased_month'] = aov['purchased_month'].astype('str')


# fig = px.bar(aov, x = 'purchased_month', y = 'total_profit (000s)', color = 'category', text = round(aov['total_profit (000s)']/1000,1) )


# #fig.update_layout(uniformtext_minsize=3, uniformtext_mode='hide')
# fig.update_traces(textfont_size=8)
# fig.show()

In [81]:
# aov = order.merge(order_prod, on = 'order_id')
# aov = aov[aov.order_status == 'delivered']
# aov = aov.merge(prod, on = 'product_id')

# per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

# aov['purchased_month'] = per
# aov['profit'] = aov['price'] - aov['shipping']
# aov = aov.groupby(['purchased_month','category'])['price','shipping','profit'].sum().reset_index()

# #aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
# aov.columns = ['purchased_month', 'category', 'total_order_item_value', 'total_shipping_cost', 'total_profit (000s)']
# aov = aov.sort_values(by = ['purchased_month','total_profit (000s)'], ascending = False)
# #aov = aov.groupby(['purchased_month']).head(3)
# aov['purchased_month'] = aov['purchased_month'].astype('str')


# fig = px.line(aov, x = 'purchased_month', y = 'total_profit (000s)', color = 'category', text = round(aov['total_profit (000s)']/1000,1) )


# #fig.update_layout(uniformtext_minsize=3, uniformtext_mode='hide')
# fig.update_traces(textfont_size=8)
# fig.show()

#### total # of orders sold by month

In [82]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per

aov = aov.groupby(['purchased_month'])['order_id'].nunique().reset_index()


#aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
aov.columns = ['purchased_month', 'orders_sold']
aov = aov.sort_values(by = ['orders_sold'], ascending = False)
#aov = aov.groupby(['purchased_month']).head(3)
aov['purchased_month'] = aov['purchased_month'].astype('str')


fig = px.bar(aov, x = 'purchased_month', y = 'orders_sold', text = 'orders_sold' )


#fig.update_layout(uniformtext_minsize=3, uniformtext_mode='hide')
fig.update_traces(textfont_size=10)
fig.show()

#### total profit generated from orders by month

In [83]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov['profit'] = aov['price'] - aov['shipping']
aov = aov.groupby(['purchased_month'])['price','shipping','profit'].sum().reset_index()

#aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
aov.columns = ['purchased_month', 'total_order_item_value', 'total_shipping_cost', 'total_profit (000s)']
aov = aov.sort_values(by = ['purchased_month','total_profit (000s)'], ascending = False)
#aov = aov.groupby(['purchased_month']).head(3)
aov['purchased_month'] = aov['purchased_month'].astype('str')


fig = px.bar(aov, x = 'purchased_month', y = 'total_profit (000s)', text = round(aov['total_profit (000s)']/1000,1) )


#fig.update_layout(uniformtext_minsize=3, uniformtext_mode='hide')
fig.update_traces(textfont_size=10)
fig.show()

<ipython-input-83-540b28cf208f>:9: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



#### total # of items sold by month

In [84]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov = aov.groupby(['purchased_month'])['price'].size().reset_index()

#aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
aov.columns = ['purchased_month', 'items_sold']
aov = aov.sort_values(by = ['items_sold'], ascending = False)
#aov = aov.groupby(['purchased_month']).head(3)
aov['purchased_month'] = aov['purchased_month'].astype('str')


fig = px.bar(aov, x = 'purchased_month', y = 'items_sold', text = 'items_sold' )


#fig.update_layout(uniformtext_minsize=3, uniformtext_mode='hide')
fig.update_traces(textfont_size=10)
fig.show()

#### general overview of items sold by category

In [85]:
aov = order.merge(order_prod, on = 'order_id')
aov = aov[aov.order_status == 'delivered']
aov = aov.merge(prod, on = 'product_id')

per = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

aov['purchased_month'] = per
aov['profit'] = aov['price'] - aov['shipping']
aov = aov.groupby(['purchased_month','category'])['price','shipping','profit'].sum().reset_index()

#aov = aov.groupby(by='order_id')['price','shipping'].sum().reset_index()
aov.columns = ['purchased_month', 'category', 'avg_order_item_value', 'avg_shipping_cost', 'avg_profit']
aov.sort_values(by = ['purchased_month','avg_order_item_value'], ascending = False).head(20)

<ipython-input-85-bd948a39d973>:9: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



purchased_month               category  avg_order_item_value  \
756         2018-08          health_beauty             119391.01   
753         2018-08              furniture             111491.45   
757         2018-08                   home              90432.19   
771         2018-08                watches              69767.06   
768         2018-08                 sports              50860.18   
736         2018-08                   auto              45032.58   
741         2018-08  computers_accessories              40052.64   
766         2018-08                 phones              39297.08   
742         2018-08     construction_tools              35099.91   
764         2018-08              perfumery              24839.74   
765         2018-08               pet_shop              24608.07   
737         2018-08                   baby              21498.46   
760         2018-08          miscellaneous              18569.75   
770         2018-08                   toys              16923.75   
755         2018-08           garden_tools              16314.70   
769         2018-08             stationery              15445.65   
761         2018-08                  music              11026.05   
745         2018-08            electronics              10739.57   
752         2018-08             food_drink              10626.05   
740         2018-08              computers               8711.50   

     avg_shipping_cost  avg_profit  
756           16253.50   103137.51  
753           27339.65    84151.80  
757           18463.18    71969.01  
771            7043.78    62723.28  
768            8810.27    42049.91  
736            7611.77    37420.81  
741            7834.22    32218.42  
766            5021.44    34275.64  
742            5200.41    29899.50  
764            3563.23    21276.51  
765            3884.39    20723.68  
737            4733.66    16764.80  
760            3156.47    15413.28  
770            3267.80    13655.95  
755            3923.39    12391.31  
769            3403.47    12042.18  
761            1278.30     9747.75  
745            2185.03     8554.54  
752            2111.15     8514.90  
740             516.75     8194.75

## customer analytics

#### repeated customer rate

This is to understand the distribution of total number of users by # of orders placed. For example, we have 3 users who placed orders 7 times. This is important analysis to understand how many "loyal" customers there are. 
From below, we learned that about 93.6% of users had cnt_order = 1. This is not a good news; only about 6.4% of users have cnt_order > 1, meaning most of customers ended up placed only 1 order, only 6.4% of entire users base are repeated customers. How can we encourage repeated behavior? We will need to identify which products each user group particularly like to encourage repeating behavior.

In [86]:
repeated_cr = customer.groupby('customer_unique_id')['customer_order_id'].nunique().reset_index()
repeated_cr.columns = ['customer_unique_id', 'cnt_order']

In [87]:
repeated_cr.groupby('cnt_order')['customer_unique_id'].size()

cnt_order
1     93099
2      2745
3       203
4        30
5         8
6         6
7         3
9         1
17        1
Name: customer_unique_id, dtype: int64

In [88]:
round( ( repeated_cr.groupby('cnt_order')['customer_unique_id'].size() / sum(repeated_cr.cnt_order) ) * 100, 2)

cnt_order
1     93.62
2      2.76
3      0.20
4      0.03
5      0.01
6      0.01
7      0.00
9      0.00
17     0.00
Name: customer_unique_id, dtype: float64

In [89]:
repeated_cr_graph = round( ( repeated_cr.groupby('cnt_order')['customer_unique_id'].size() / sum(repeated_cr.cnt_order) ) * 100, 2).reset_index()

In [90]:
repeated_cr_graph.columns = ['cnt_order', '# of customers (%)']

In [91]:
fig = px.bar(repeated_cr_graph, 
             x = 'cnt_order', 
             y = '# of customers (%)')
fig.show()

## operational analytics

* understand which category has higher cancel rate
    * highest category was home
* get avg. latency rate and avg. shipping cost for products that are delivered by category
    * there was no strong relationship between shipping cost and latency rate

In [92]:
df = order.merge(order_prod, on = 'order_id')
df = df.merge(prod, on ='product_id')    

In [93]:
df_status = df.groupby(['order_status','category'])[['order_id']].nunique().reset_index()

In [94]:
df_status.columns = ['order_status','category','order_cnt']

In [95]:
oper = order.merge(order_prod, on = 'order_id')
oper.groupby(['order_status'])['order_id'].nunique()

order_status
approved           2
canceled         461
delivered      96478
invoiced         312
processing       301
shipped         1106
unavailable        6
Name: order_id, dtype: int64

In [96]:
df_top_5 = df_status.sort_values(by='order_cnt', ascending = False).groupby('order_status').head(5).sort_values(by = ['order_status', 'order_cnt'])

In [97]:
fig = px.bar(df_top_5[df_top_5['order_status'] == 'canceled'], 
             x = 'category', 
             y = 'order_cnt')
fig.show()

In [98]:
oper = order.merge(order_prod, on = 'order_id')
oper['latency_day'] = pd.to_datetime(oper['delivered_at']) - pd.to_datetime(oper['purchased_at'])
oper['latency_day'] = oper.latency_day.dt.days
oper['profit'] = oper['price'] - oper['shipping']
oper.describe()

num_items_in_order          price       shipping    latency_day  \
count       112650.000000  112650.000000  112650.000000  110196.000000   
mean             1.197834     120.653739      19.990320      12.007723   
std              0.705124     183.633928      15.806405       9.451455   
min              1.000000       0.850000       0.000000       0.000000   
25%              1.000000      39.900000      13.080000       6.000000   
50%              1.000000      74.990000      16.260000      10.000000   
75%              1.000000     134.900000      21.150000      15.000000   
max             21.000000    6735.000000     409.680000     209.000000   

              profit  
count  112650.000000  
mean      100.663419  
std       177.670283  
min      -140.530000  
25%        24.520000  
50%        56.290000  
75%       111.830000  
max      6540.690000

In [99]:
oper = oper.groupby(['latency_day'])['shipping'].mean().reset_index()
oper

latency_day   shipping
0            0.0  16.211111
1            1.0  10.623539
2            2.0  11.523158
3            3.0  13.282303
4            4.0  14.182021
..           ...        ...
141        191.0  54.330000
142        194.0  53.606667
143        195.0  25.120000
144        208.0  17.260000
145        209.0  15.780000

[146 rows x 2 columns]

In [100]:
#oper = pd.DatetimeIndex(aov.purchased_at).to_period("M") 

#oper['purchased_month'] = per

#oper = oper.groupby(['latency_day'])['price'].mean().reset_index()

#oper['purchased_month'] = oper['purchased_month'].astype('str')
oper.columns = ['avg. latency_day', 'avg. shipping cost']
fig = px.scatter(oper, x = 'avg. latency_day', y = 'avg. shipping cost')
fig.show()

## Summary

* We see that furniture generates highest profit for Bon (total profit of 1.47m dollars, avg. profit of 73 dollars per item sold, total 200k items being sold).

* Keep in mind that watches are 2nd largest profit generator (total profit of 1.1m dollars which is about 0.37m dollars lower than furniture and its not that much of difference, avg. profit of 182 dollars per item sold, total 5.9k items being sold) with more than 2 times higher avg. profit per item sold than furniture. 

* By examining total profit/avg. profit per item/total items sold, we can examine which items we should focus more on to increase the revenue in the future.

* We can say watches have potnetial to become the largest profit generator in the future if we promote more on these items as it has higher profit than furniture. It will be important to send out more flyers on watches and possibly give some discount on these goods at times. Since it yields high avg. profit per item sold (top 4), Bonsai might be able to diversify portfolios to reduce risk of fully depending on just furniture as main driver of profit generator (large quantity but low/medium avg. profit such as furniture vs low/medium quantity but high avg.profit such as watches)

* We see that Bon's total profit is generally in decline since June 2018. The main driver of decline would be decline in general sales of top 4 categories. This is another reason to diversify product portfolios - keep in mind that computers and computer accessories used to be fairly high in profit but dramatically decreased for some reason. We will need to find out reasons why this happened and try hard to increase the sales from these to diversify our product portfolio.

* From monthly timeseries total profit plot, I discovered that categories such as watches and health product eventually overtook furniture (which used to be the largest profit generator almost every month) in terms of total monthly profit since May (watches became largest profit generator in May 2018 with 111k in dollar and health beauty eventually took over watches to claim the largest profit generator in Aug 2018 with total profit of 103k in dollar).

* We will have to understand why the change happend - is it because we gave more promotion on health beauty product or did consumer's preference just changed? We need to identify this with product team. If it is generally due to the change in preference, I would recommend product/marketing team to place more health beauty products on partner's websites. If we think this might be due to promotion impact then we will need to run A/B test to find if this is due to random chance or not.

* From average order value (and order item value)/shipping cost/profit per order (and per item) sold by month graph, we see that avg. profit per order/item had been in decline. This is a little bit of yellow flag for Bon as users place less orders (and also buy less items) especially in Aug 2018 -- refer to 2 graphs. How can we effectively increase average order value (and order item value)? We can encourage users buy more watches on discount or find customers who usually buy higher valued goods on average to encourage them continue their behaviors.